spring-integration-zookeeper/src/main/java/org/springframework/integration/zookeeper/lock/ZookeeperLockRegistry.java

```diff
@@ -0,0 +1,261 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.integration.zookeeper.lock;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.recipes.locks.InterProcessMutex;
+
+import org.springframework.integration.support.locks.ExpirableLockRegistry;
+import org.springframework.messaging.MessagingException;
+import org.springframework.util.Assert;
+
+/**
+ * {@link ExpirableLockRegistry} implementation using Zookeeper, or more specifically,
+ * Curator {@link InterProcessMutex}.
+ *
+ * @author Gary Russell
+ * @since 4.2
+ *
+ */
+public class ZookeeperLockRegistry implements ExpirableLockRegistry {
+
+	private static final String DEFAULT_ROOT = "/SpringIntegration-LockRegistry";
+
+	private final CuratorFramework client;
+
+	private final KeyToPathStrategy keyToPath;
+
+	private final Map<String, ZkLock> locks = new HashMap<String, ZkLock>();
+
+	private final boolean trackingTime;
+
+	/**
+	 * Construct a lock registry using the default {@link KeyToPathStrategy} which
+	 * simple appends the key to '/SpringIntegration-LockRegistry/'.
+	 * @param client the {@link CuratorFramework}.
+	 */
+	public ZookeeperLockRegistry(CuratorFramework client) {
+		this(client, DEFAULT_ROOT);
+	}
+
+	/**
+	 * Construct a lock registry using the default {@link KeyToPathStrategy} which
+	 * simple appends the key to {@code '<root>/'}.
+	 * @param client the {@link CuratorFramework}.
+	 * @param path the path root (no trailing /).
+	 */
+	public ZookeeperLockRegistry(CuratorFramework client, String root) {
+		this(client, new DefaultKeyToPathStrategy(root));
+	}
+
+	/**
+	 * Construct a lock registry using the supplied {@link KeyToPathStrategy}.
+	 * @param client the {@link CuratorFramework}.
+	 * @param keyToPath the implementation of {@link KeyToPathStrategy}.
+	 */
+	public ZookeeperLockRegistry(CuratorFramework client, KeyToPathStrategy keyToPath) {
+		Assert.notNull(client, "'client' cannot be null");
+		Assert.notNull(client, "'keyToPath' cannot be null");
+		this.client = client;
+		this.keyToPath = keyToPath;
+		this.trackingTime = !keyToPath.bounded();
+	}
+
+	@Override
+	public Lock obtain(Object lockKey) {
+		Assert.isInstanceOf(String.class, lockKey);
+		String path = this.keyToPath.pathFor((String) lockKey);
+		ZkLock lock = this.locks.get(path);
+		synchronized (this.locks) {
+			lock = this.locks.get(path);
```

Comment:
Why is `this.locks.get(path)` twice ?
I'm not sure in the algorithm, but I feel that everything around `this.locks` must be done within `synchronized`